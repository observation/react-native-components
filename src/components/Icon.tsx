import React from 'react'

import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'

import Icons, { IconName } from '../lib/Icons'
import theme from '../styles/theme'

export type IconStyleProp = 'light' | 'solid'

type IconAppearanceProps = {
  style?: IconStyleProp
  color?: string
  size?: number
  testID?: string
}

type IconProps = IconAppearanceProps & {
  name: IconName
}

export const Icon = ({ name, color, size, testID, style }: IconProps): JSX.Element => {
  const iconStyle = style ?? 'light'
  const icon = iconStyle === 'light' ? Icons[name].light : Icons[name].solid
  const iconColor = color ?? theme.color.primary
  const iconSize = size ?? theme.icon.size.large

  return <FontAwesomeIcon icon={icon} color={iconColor} size={iconSize} testID={testID} />
}
