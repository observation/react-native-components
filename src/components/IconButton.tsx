import React from 'react'
import { TouchableOpacity, ViewStyle, StyleProp } from 'react-native'

import { Icon, IconProps } from './Icon'
import theme from '../styles/theme'

type Props = {
  containerStyle?: StyleProp<ViewStyle>
  disabled?: boolean
  onPress?: () => void
  icon: IconProps
  accessibilityLabel?: string
  testID?: string
}

const IconButton = ({ containerStyle, disabled, onPress, icon, accessibilityLabel, testID = 'pressable' }: Props) => (
  <TouchableOpacity
    testID={testID}
    accessibilityLabel={accessibilityLabel}
    style={containerStyle}
    disabled={disabled}
    onPress={disabled ? undefined : onPress}
    activeOpacity={0.5}
  >
    <Icon size={theme.icon.size.large} {...icon} />
  </TouchableOpacity>
)

export default IconButton
