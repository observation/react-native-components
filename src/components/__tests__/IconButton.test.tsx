import React from 'react'

import { render, fireEvent } from '@testing-library/react-native'

import { IconProps } from '../Icon'
import IconButton from '../IconButton'

describe('IconButton', () => {
  const icon: IconProps = { name: 'chevron-left', size: 20, color: 'black' }
  test('Rendering', () => {
    // GIVEN
    const { toJSON } = render(<IconButton containerStyle={{ flex: 1 }} disabled={false} icon={icon} />)

    // THEN
    expect(toJSON()).toMatchSnapshot()
  })

  test('Handle press when enabled', () => {
    // GIVEN
    const mockOnPress = jest.fn()
    const { getByTestId } = render(<IconButton disabled={false} onPress={mockOnPress} icon={icon} />)

    // WHEN
    const pressable = getByTestId('pressable')
    fireEvent.press(pressable)

    // THEN
    expect(mockOnPress).toHaveBeenCalled()
  })

  test('Do not handle press when disabled', () => {
    // GIVEN
    const mockOnPress = jest.fn()
    const { getByTestId } = render(<IconButton disabled onPress={mockOnPress} icon={icon} />)

    // WHEN
    const pressable = getByTestId('pressable')
    fireEvent.press(pressable)

    // THEN
    expect(mockOnPress).toHaveBeenCalledTimes(0)
  })
})
