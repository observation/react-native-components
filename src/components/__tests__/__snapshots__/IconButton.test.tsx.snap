// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton Rendering 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flex": 1,
      "opacity": 1,
    }
  }
  testID="pressable"
>
  <Icon
    color="black"
    name="chevron-left"
    size={20}
    type="light"
  />
</View>
`;
